### PR TITLE
Support running PreCommitCondition#cleanup before Transaction#onSuccess where possible.

### PR DESCRIPTION
**Pull request by mdaudali**
## General
**Before this PR**:
PreCommitConditions can be used to hold exclusive locks, released during cleanup. AtlasDB explicitly mentions locks within the precommit condition docs.
Transactions support `onSuccess` callbacks, that may run once the transaction has been successfully committed.

However, the onSuccess callbacks run _before_ PreCommitCondition cleanup, which can lead to lock contention and deadlocking.

After an internal review, we decided that we should guarantee that cleanup runs before onSuccess under certain conditions (namely, commit is not called within the transaction task).

**After this PR**:

==COMMIT_MSG==
If the cleanup is to run, `PreCommitCondition#cleanup()` is guaranteed to run before `Transaction#onSuccess()` callbacks, provided that `Transaction#commit()` was not called manually.
==COMMIT_MSG==

**Priority**: P1

**Concerns / possible downsides (what feedback would you like?)**:
Is there a cleaner way? This avoids touching Transaction but the CallbackAwareTransaction is a little bit intrusive.
There's probably a much better name than CallbackAwareTransaction...

Correctness!

Whether we want to support onSuccess running at most once (we decided internally not, but whether in the face of it we review that)
**Is documentation needed?**:
The interface comments have been updated
## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:
No
**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:
No
**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:
Yes
**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:
No
**Does this PR need a schema migration?**
No
## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:
That the touched places are the only places that run condition cleanup.
That defining this ordering will not break anyone in a way we're not okay with.
**What was existing testing like? What have you done to improve it?**:
Removed test that guarantees onSuccess running at most once 
Added tests to show that cleanup doesn't prevent committing, and cleanup runs before onSuccess.
**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:
N/A
**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:
N/A, we continue to maintain that onSuccess and Cleanup are best effort
## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:
Any clean up tasks run before onsuccess where the ordering matters
**Has the safety of all log arguments been decided correctly?**:
No extra logs.
**Will this change significantly affect our spending on metrics or logs?**:
N/A
**How would I tell that this PR does not work in production? (monitors, etc.)**:
Clean up tasks run after onSuccess where ordering matters
**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:
Rollback
**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:
N/A
## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:
No
**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:
No
**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:
No
## Development Process
**Where should we start reviewing?**:
SnapshotTransaction#commitWithoutCallback
**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:
N/A
**Please tag any other people who should be aware of this PR**:
@jeremyk-91

### CODE REVIEW

File: atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionManager.java [hunk 1]
```diff
@@ -61,6 +61,30 @@ protected final <T, E extends Exception> T runTaskThrowOnConflict(TransactionTas
         }
     }
 
+    protected final <T, E extends Exception> T runTaskThrowOnConflictWithCallback(
```
Comment (mdaudali) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionManager.java:64]: There is a method above this with the same implementation except without commitWithoutCallbacks. However, that's not used within AtlasDB, and have tried searching for uses internally (but not totally convinced that the search was sufficient). I found no uses (as this is different from runTaskThrowOnConflict(TransactionTask<T, E> task), as this has an extra parameter.

If we do want to keep both, we _could_ deduplicate by factoring out the commit method. (although that may make this less readable)

File: atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/ForwardingCallbackAwareTransaction.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.impl;
+
+import com.palantir.atlasdb.transaction.api.TransactionFailedException;
+import com.palantir.atlasdb.transaction.service.TransactionService;
+
+public abstract class ForwardingCallbackAwareTransaction extends ForwardingTransaction
```
Comment (mdaudali) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/ForwardingCallbackAwareTransaction.java:22]: This is for ReadTransaction and GetAsyncCallbackAwareDelegate.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java [hunk 3]
```diff
@@ -170,13 +170,10 @@ protected boolean shouldStopRetrying(int numTimesFailed) {
...
-        try {
-            OpenTransaction openTransaction = runTimed(
-                    () -> Iterables.getOnlyElement(startTransactions(ImmutableList.of(condition))), "setupTask");
-            return openTransaction.finish(transaction -> task.execute(transaction, condition));
-        } finally {
-            condition.cleanup();
-        }
+        OpenTransaction openTransaction =
+                runTimed(() -> Iterables.getOnlyElement(startTransactions(ImmutableList.of(condition))), "setupTask");
+        return openTransaction.finishWithCallback(
```
Comment (mdaudali) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java:175]: This is different - previously, if the openTransaction L174 threw, cleanup was called. In the new, this is not the case. We can solve this by something like 

```
 checkOpen();
        OpenTransaction openTransaction;
        try {
            openTransaction = runTimed(
                    () -> Iterables.getOnlyElement(startTransactions(ImmutableList.of(condition))), "setupTask");
        } catch (Exception e) {
            condition.cleanup();
            throw e;
        }
        return openTransaction.finishWithCallback(
                transaction -> task.execute(transaction, condition), condition::cleanup);
```

at the expense of being less readable. We do not explicitly guarantee that cleanup will run

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/GetAsyncCallbackAwareDelegate.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+package com.palantir.atlasdb.transaction.impl;
+
+import com.palantir.atlasdb.futures.AtlasFutures;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import java.util.Map;
+import java.util.Set;
+
+public class GetAsyncCallbackAwareDelegate extends ForwardingCallbackAwareTransaction {
```
Comment (mdaudali) [inline atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/GetAsyncCallbackAwareDelegate.java:25]: Test utils, expands on `GetAsync` for SnapshotTransaction tests.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 5]
```diff
@@ -1077,6 +1077,34 @@ public void cleanup() {
...
+            return null;
+        });
+        assertThat(counter.intValue()).isEqualTo(2);
+
+        counter.setValue(0);
+        serializableTxManager.runTaskWithConditionReadOnly(latchingCondition, (tx, condition) -> {
+            tx.onSuccess(() -> assertThat(counter.getAndAdd(1)).isEqualTo(1));
+            return tx.get(TABLE, ImmutableSet.of(TEST_CELL));
+        });
+        assertThat(counter.intValue()).isEqualTo(2);
```
Comment (jeremyk-91) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1105]: This is compact, but would suggest for readability to have a separate test for the rTWCTOC and rTWCRO cases, and ideally use a separate counter/condition in both; the reset halfway was a bit confusing

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 6]
```diff
@@ -1904,6 +1921,31 @@ public void transactionStillCommittedEvenIfCallbackThrows() {
...
+                throw new RuntimeException("boom");
+            }
+        };
+
+        assertThatThrownBy(() -> txManager.runTaskWithConditionThrowOnConflict(preCommitCondition, (txn, condition) -> {
+                    txn.put(TABLE, ImmutableMap.of(TEST_CELL, PtBytes.toBytes("tom")));
+                    return null;
+                }))
+                .isInstanceOf(RuntimeException.class)
+                .hasMessageContaining("boom");
```
Comment (jeremyk-91) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1941]: I normally prefer to have an exception declared as a constant and reference it both in your preCommitCondition and assertion (that they're equal)

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java (hunk 3 above)
Comment (jeremyk-91) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java:175]: Yeah, we don't. It is best effort, but I think we should (and what you describe looks right): startTransactions does involve an RPC to TimeLock, so there's a pretty real probability that that fails.

**Review by jeremyk-91 — COMMENTED** [review]
Broadly looks good. Don't really have a better design :(

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 6 above)
Comment (mdaudali) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1941]: Yeah that's reasonable - I admit I copied the onSuccess version of this directly above

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 7]
```diff
@@ -1923,29 +1917,40 @@ public void transactionStillCommittedEvenIfCallbackThrows() {
...
+                .hasMessageContaining(exception.getMessage());
         txManager.runTaskReadOnly(txn -> {
             assertThat(txn.get(TABLE, ImmutableSet.of(TEST_CELL)))
                     .containsExactly(Maps.immutableEntry(TEST_CELL, PtBytes.toBytes("tom")));
             return null;
         });
     }
 
+    @Test
+    public void preCommitCleanupHappensIfStartTransactionsFails() {
```
Comment (jeremyk-91) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1939]: 👍

**Review by jeremyk-91 — APPROVED** [review]
👍  Thanks!

**Pull request merged**